cmake/CompilerChecks.cmake

```diff
@@ -23,11 +23,65 @@ if(NOT DEFINED CMAKE_INTERPROCEDURAL_OPTIMIZATION)
     endif()
 endif()
 
+
 # Check for CPUID headers
 try_compile(HAVE_CPUID ${PROJECT_BINARY_DIR}
             ${PROJECT_SOURCE_DIR}/cmake/check_cpuid.cpp)
-if(HAVE_CPUID)
-    message(STATUS "CPUID header found -- using hardware math acceleration when available")
-else()
-    message(STATUS "CPUID header not found -- using software math")
-endif()
+
+# Check for SIMD headers
+CHECK_INCLUDE_FILE("immintrin.h" HAVE_AVX2)
+CHECK_INCLUDE_FILE("immintrin.h" HAVE_AVX)
+CHECK_INCLUDE_FILE("nmmintrin.h" HAVE_SSE42)
+CHECK_INCLUDE_FILE("tmmintrin.h" HAVE_SSSE3)
+CHECK_INCLUDE_FILE("smmintrin.h" HAVE_SSE41)
+CHECK_INCLUDE_FILE("tmmintrin.h" HAVE_SSE4)
+CHECK_INCLUDE_FILE("pmmintrin.h" HAVE_SSE3)
+CHECK_INCLUDE_FILE("emmintrin.h" HAVE_SSE2)
+CHECK_INCLUDE_FILE("xmmintrin.h" HAVE_SSE1)
+
+# GCC requires us to set the -m<instructionset> flag for the source file using the intrinsics.
+# We can't do that project-wide or we'll just crash on launch with an illegal instruction on some
+# systems. So, we have another helper method...
+function(plasma_target_simd_sources TARGET)
+    set(_INSTRUCTION_SETS "SSE1;SSE2;SSE3;SSE4;SSE41;SSSE3;SSE42;AVX;AVX2")
+    set(_GCC_ARGS "-msse;-msse2;-msse3;-msse4;-msse4.1;-mssse3;-msse4.2;-mavx;-mavx2")
+    cmake_parse_arguments(PARSE_ARGV 1 _passf "" "SOURCE_GROUP" "${_INSTRUCTION_SETS}")
+
+    # Hack: if we ever bump to CMake 3.17, use ZIP_LISTS.
+    list(LENGTH _INSTRUCTION_SETS _LENGTH)
+    math(EXPR _LENGTH "${_LENGTH} - 1")
+
+    foreach(i RANGE ${_LENGTH})
+        # foreach(i IN ZIP_LISTS _INSTRUCTION_SETS _GCC_ARGS)
+        list(GET _INSTRUCTION_SETS ${i} _CURRENT_INSTRUCTION_SET)
+        list(GET _GCC_ARGS ${i} _CURRENT_COMPILE_OPTION)
+
+        set(_CURRENT_SOURCE_FILES ${_passf_${_CURRENT_INSTRUCTION_SET}})
+        if(_CURRENT_SOURCE_FILES)
+            target_sources(${TARGET} PRIVATE ${_CURRENT_SOURCE_FILES})
+            if(_passf_SOURCE_GROUP)
+                source_group(${_passf_SOURCE_GROUP} FILES ${_CURRENT_SOURCE_FILES})
+            endif()
+            if(HAVE_${_CURRENT_INSTRUCTION_SET} AND CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
+                set_source_files_properties(
+                    ${_CURRENT_SOURCE_FILES} PROPERTIES
+                    COMPILE_OPTIONS ${_CURRENT_COMPILE_OPTION}
```

Comment:
Does this replace the property or add to it?

Reply:
Replace, which is a bad thing, I suppose...